### PR TITLE
Solution for CF-963D

### PR DESCRIPTION
**Pull request by dutinmeow**

### CODE REVIEW

**Review by nathangong — APPROVED** [review]

**Comment by dutinmeow** [comment]
> you know what a constructor is
why not actually
use it

Using an init/build function is pretty normal in CP ([example](https://github.com/bqi343/USACO/blob/master/Implementations/content/data-structures/1D%20Range%20Queries%20(9.2)/SegTree%20(9.2).h)). Anyhow, I don't feel like it makes the code any more confusing or unclear

**Comment by SansPapyrus683** [comment]
... ok but why don't you want to put it in a constructor?

File: solutions/cf-963D.mdx
Comment (SansPapyrus683) [inline solutions/cf-963D.mdx:50]:
```suggestion
```
why does this still exist lol

File: solutions/cf-963D.mdx
Comment (SansPapyrus683) [inline solutions/cf-963D.mdx:108]: does it run w/o the +1?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by dutinmeow** [comment]
> does it run w/o the +1?
You yourself said it was one-based

**Comment by dutinmeow** [comment]
> why does this still exist lol

A default constructor is necessary because I declare the Suffix Automaton in the global scope before the string is inputted. It would be a syntax error if it wasn't there (or I could put the construct method in an `init` function...)

**Comment by SansPapyrus683** [comment]
i'll just believe you
will review later, ping me if i forget

**Comment by dutinmeow** [comment]
how much later

File: solutions/cf-963D.mdx
Comment (Aadit-Ambadkar) [reply solutions/cf-963D.mdx:50]: you need a constructor...

File: solutions/cf-963D.mdx
Comment (Aadit-Ambadkar) [inline solutions/cf-963D.mdx:115]:
```suggestion
// answer for query
int Q, ans[MAXN];  
// queries per state
vector<pii> queries[MAXN];  
// suffix link tree (at most 2 * |S| nodes)
vector<int> suffix_link_tree[MAXN * 2]; 
```

The last line was over 80 characters.

File: solutions/cf-963D.mdx
Comment (Aadit-Ambadkar) [inline solutions/cf-963D.mdx:119]:
```suggestion
void dfs(int u) {
```

File: solutions/cf-963D.mdx
Comment (Aadit-Ambadkar) [inline solutions/cf-963D.mdx:165]:
```suggestion
	dfs(0);
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/cf-963D.mdx
Comment (Aadit-Ambadkar) [reply solutions/cf-963D.mdx:50]: You still need a constructor

Comment (SansPapyrus683) [reply solutions/cf-963D.mdx:50]: no he declared one @ the bottom
dustin explained in one of his comments

**Review by Aadit-Ambadkar — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
i'll wait for someone who actually knows about suffixes to review before i merge

**Pull request merged**
